### ISSUE TITLE
(OUTPUT 2) Must support output of alignment as self-contained HTML with the following features

### ISSUE THREAD

**Issue by irina060981** [alpheios-project/alignment-editor-new#70]
Must support output of alignment as self-contained HTML with the following features
- mouseover/interlinear display of alignments as in the editing interface (but read only)
- any line breaks shown
- several display modes:
-- full panel on right devoted to a single translation (as now) chosen by the user, but allowing the user to change the translation at will
-- right hand scrollable panel displays only the words that match the selected words in the original on the left in a scrollable panel- from as many translations as the user has chosen and in the order they have chosen
-- right hand scrollable panel displays only the complete sentences containing any words matching the word or phrase selected in the origina l- from as many translations as the user has chosen and in the order they have chosen. (An even broader context might be considered, eg a sentence before and after).
- the ability of the user to toggle among the different display modes while viewing the alignment
- if it contains multiple translations, the user should be able to choose the sequence in which they are displayed and how many to display at once
- the ability to select different translations for different sections of the text
- an option to embed the Alpheios Reading Tools in the display

From https://github.com/alpheios-project/documentation/blob/master/development/alignment-editor.md

**Comment by irina060981** [comment]
The frst step for OUTPUT (discussed on the meeting with @balmas and @abrasax :

from Slack:

Screenshots with proposals:
![Alignment Editor Modes Output](https://user-images.githubusercontent.com/12377640/105677544-b8035c00-5f37-11eb-9468-3fd43246eee3.png)

**@irina060981:** I believe that the output would have two parts - some big rich menu and the result as displayed on the picture
in the menu - we would choose translations, orders for each segment (I believe section === segment in the description)

**@balmas:** yes, that's right. and yes, re section/segment. It's a tricky word -- sentence/section/segment so it can lead to some confusion but I tried to use the general term segment to mean whatever syntactic divisions are important to the user

**@irina060981:** it should work as a local file

**@balmas:** yes, that's a good point to keep in mind, because we need to be able to provide the user with a standalone version of the display

**@irina060981:** I only worry would alpheios embed-library work inside local file

**@balmas:** it does actually work in local files, mostly some features, such as the treebank disambiguation, does not
and it doesn't work locally in safari  on mac (I think, definitely there is at least one browser on mac it doesn't work locally in)
but the alpheios embedded library functioning working in the standalone local display is less critical than the alignment functionality working

**Comment by irina060981** [comment]
From Slack

**@balmas:**
This was Harry's response to those screenshots:
the full origin text should always display, only the translations should be limited
display options:
1            full half page on either side exactly as in original editor
             with word or words on the original text side linked to corresponding
             word or words on the translation side
   left side          right side:
2  word or words    word x
   linked to x        metadata from translation  (translator, date, title, publisher, etc ONLY ONE LINE))
   word or words    word y
   linked to y        metadata from translation b
   word or words    word x
   linked to x        metadata from translation c
   word or words    word z "..." k
   linked to z and k  metadata from translation d
   (one concept-ie
   one link)
             -----------------------------------
3    "        sentence start [word x] sentence end.
             metadata from translation a (translator, date, title, publisher, etc ONLY ONE LINE)
              sentence start [word y] sentence end.
             metadata from translation b
              sentence start [word x] sentence end.
             metadata from translation c
              sentence start [word z] sentence continuation [word k] sentence end.
             metadata from translation d
             sentence start [word a] sentence end.
              sentence start [word aa] sentence end.
             metadata from translation e
             (rare case where match requires more than one sentence)

**Comment by irina060981** [comment]
From Slack:

**@balmas:**
a quick note about sentence. We will need to be able to isolate sentences within a segment.  My intent, if we wanted to support this, was to have the tokenizer identify the sentences at tokenization time and include that information in the token metadata. This won't be possible to do right now, so for the immediate term identification of sentences will have to be done client side. The tokenizer DOES identify punctuation, via the boolean punct property so you should be able to use that, although it's not limited to sentence-ending punctuation (e.g. it will identify quotes as punct=true so  you'll have to filter non-sentence break punctuation out.

**Comment by irina060981** [comment]
View from @abrasax :

[displayOption2.pdf](https://github.com/alpheios-project/alignment-editor-new/files/5865123/displayOption2.pdf)

**Comment by irina060981** [comment]
From the meeting (21 Jan 2021):

- All text parts should be scrolled independently
- when tokens (from a group) a hovered in one text - in other texts tokens from the group should be scrolled to be visible
- should add a sentence index - to show one o more sentences around the word

Decided first solve the issue https://github.com/alpheios-project/alignment-editor-new/issues/124
Then Harry will prepare test texts
then do this issue

**Comment by irina060981** [comment]
The first step is done.
Features for a user:
  - added download type - Html
  - created options 1 - the same layout as in editor
  - the same theme is applied as defined on download
  - limited height for the segments block and added independent scroll
  - auto scroll to the invisible part of hovered alignment group
  
![image](https://user-images.githubusercontent.com/12377640/107037107-5a5bf300-6806-11eb-90a2-9a42b08e7b74.png)

![image](https://user-images.githubusercontent.com/12377640/107037306-9c853480-6806-11eb-8600-944142d79621.png)

**Comment by irina060981** [comment]
The structure specific.
I decided to do this the following way:
- I created a separate folder in `scr` folder for `_output`
- I have styles and Vue components for the creating Output HTML in th `_output` folder that are independent from the other code
- I have a separate `config-output.mjs` for building styles and Vue with output components 
![image](https://user-images.githubusercontent.com/12377640/107037808-567ca080-6807-11eb-9cbc-54ef1fece13a.png)

- I created an `html-template.json` for creating html dwonload in template manner.
Here it is the structure:

```
{
  "params": [ "theme", "stylePath", "jsPath", "pageTitle", "fullData" ],
  "stylePath": "file:///C:/_Alpheios/alignment-editor/public/dist/output/style/style-alignment-editor-output.css",
  "jsPath": "file:///C:/_Alpheios/alignment-editor/public/dist/output/alpheios-alignment-editor-output.js",
  "pageTitle": "Alpheios Alignment Editor|Result",
  "layout": "
       <!DOCTYPE html> 
       <html class=\"{{theme}}\"> 
       <head> 
              <meta charset=\"UTF-8\"> 
              <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"> 
              <title>{{pageTitle}}</title> 
              <link rel=\"icon\" type=\"image/png\" href=\"https://alignment.alpheios.net/logo.png\"> 
              <link rel=\"stylesheet\" href=\"{{stylePath}}\"/> 
              <script src=\"{{jsPath}}\"></script>
       </head> 
       <body class=\"{{theme}}\"> 
              <div class=\"container\"> 
                     <div id=\"alpheios-alignment-editor-output\"></div> 
              </div> 
              <script> 
                     document.addEventListener(\"DOMContentLoaded\", function(event) { 
                            new window.AlignmentEditorOutput.Vue({ 
                                   data: { fullData: {{fullData}} }, 
                                   render: (h) => h(window.AlignmentEditorOutput.App)
                            }).$mount('#alpheios-alignment-editor-output') }) 
              </script>
       </body> 
</html>"
}
```

And on download preparation step I replace template tags `{{}}` with values.
`FullData` - is a specific conversion of alignment data - with all tokens and groups.
Has the following structure (an example part)
```
{
  "origin": {
    "dir": "ltr",
    "lang": "lat",
    "segments": [
      {
        "tokens": [
          {
            "textType": "origin",
            "idWord": "1-0-0",
            "word": "Capuam",
            "grouped": true,
            "groupData": [
              {
                "groupId": "dbc24e70-e7c5-4e6a-b4f4-c16eb6826cb7",
                "targetId": "4c46e745-ebfc-48ec-85ad-173c26d87c54"
              }
            ]
          },
```

**Comment by irina060981** [comment]
For now it is attached to the local files - after upload to AWS - I would change style and js libraries paths to the AWS

**Comment by irina060981** [comment]
The second step is done

- added radio select - view types
![image](https://user-images.githubusercontent.com/12377640/107358570-47576480-6b1f-11eb-9eba-924db552edcc.png)

- Full view - is from display option 1:
```
full half page on either side exactly as in original editor
with word or words on the original text side linked to corresponding
word or words on the translation side
```
- Short view - is from display option 2:
![image](https://user-images.githubusercontent.com/12377640/107358783-8a193c80-6b1f-11eb-9470-38e170e4db5f.png)

**Comment by irina060981** [comment]
Added display option 3:
![image](https://user-images.githubusercontent.com/12377640/107510351-8359fa80-6bef-11eb-93e7-5508a726b75a.png)

- added sentence around input to define, how much sentences to show around (0 - only current, 1 - one before, one after and so on)
![image](https://user-images.githubusercontent.com/12377640/107510425-a5ec1380-6bef-11eb-90d2-58269ac24847.png)

**Comment by irina060981** [comment]
@monzug - I think this issue is ready for your testing and verification

**Comment by abrasax** [comment]
Does "points" include  ? and ! ?  We might actually want to include : and
;  since they typically mark semantic unitss.

**Comment by monzug** [comment]
Thanks. will look into it.

**Comment by abrasax** [comment]
thanks,  I should have an opportunity today.

**Comment by irina060981** [comment]
>Does "points" include  ? and ! ?  We might actually want to include : and
;  since they typically mark semantic unitss.

thank you, @abrasax , - will add them!

**Comment by balmas** [comment]
@irina060981 note that there is the is_punct metadata on the token that you could check for this.  The tokenizer identifies language-specific punctuation.

**Comment by balmas** [comment]
(of course it's not just end-of sentence punctuation, but you could filter on that at least)

**Comment by irina060981** [comment]
@balmas , yes, I use this flag, but it shows true for ',' and other punctuation, so I need additional filter by specific punctuation.
for now it is - ['.', ';', '!', '?', ':']

we could expand it while testing.

But I have a question - is line break an end of the sentence ?

**Comment by abrasax** [comment]
It's nice to have the option to expand the window around the target
sentence (or sentences if matches occur in more than one sentence at a
time) as a user option- but the most important version is to just display
the sentence (or sentences) in which an actual match occurs. And that
should be the default.  Remember that the object is to be able to show as
many alternative translations in the smallest space.  So we really have
only three basic displays- the complete text,  the sentence (or sentences)
in which a match occurs, and just the matched items themselves (with no
sentence context)- together with a condensed metadata line. We could make
the expanding window around the sentence a fourth option, but I am now
tempted to leave that under a general Options menu, or perhaps a Display
Options menu.

**Comment by abrasax** [comment]
line break is not an end of sentence for this purpose.

**Comment by irina060981** [comment]
> line break is not an end of sentence for this purpose.
ok

**Comment by abrasax** [comment]
I see it's pretty late in Vladivostok- perhaps we should continue in the
morning.

**Comment by irina060981** [comment]
ok - I will read everything on my yesterday morning :)

**Comment by monzug** [comment]
1) the download as html is possible after alignment. Would it be possible to disable the html option (greyed out or not visible) before tokenization process has started?
2) super minor thing: alpheios icon is missing in Safari tab (it's visible in chrome and firefox)
![Screen Shot 2021-02-10 at 2 58 03 PM](https://user-images.githubusercontent.com/41396793/107523664-44a74e80-6bb5-11eb-935f-4b9ad55375c6.png)
3) when I open the html file, it defaults to sentence view. are we ok with this?
4) sentence view: in the translated text, I am missing the edited annotation which is visible in Full view. why is that? see attachment
![Screen Shot 2021-02-10 at 3 57 55 PM](https://user-images.githubusercontent.com/41396793/107527299-fc8a2b00-6bb8-11eb-9d88-90eeb6cc0cf4.png)
5) sentence view is supposed to show how many sentences to show around (0 - only current, 1 - one before, one after and so on). Going up works fine, however going down didn't work for me. I was expecting to see the sentences below the current selection but it did only show the word selected and the related alignment. see attachments
![Screen Shot 2021-02-10 at 4 15 00 PM](https://user-images.githubusercontent.com/41396793/107529681-69062980-6bbb-11eb-8d9c-2e7b5d05c1d4.png)
![Screen Shot 2021-02-10 at 4 15 16 PM](https://user-images.githubusercontent.com/41396793/107529713-715e6480-6bbb-11eb-913e-82bfd87bf9a0.png)
6) the language is not relevant in html output. Please confirm that this is ok.

**Comment by irina060981** [comment]
>the download as html is possible after alignment. Would it be possible to disable the html option (greyed out or not visible) before tokenization process has started?

will do

**Comment by irina060981** [comment]
>super minor thing: alpheios icon is missing in Safari tab (it's visible in chrome and firefox)

will check

**Comment by irina060981** [comment]
> when I open the html file, it defaults to sentence view. are we ok with this?

I don't know what item should be the default - I made it as sentence only because it was comfotable for me for developing

**Comment by irina060981** [comment]
>sentence view: in the translated text, I am missing the edited annotation which is visible in Full view. why is that? see attachment

Monica, I didn't understand your comment. What is edited annotation?
Could you attach both - full and sentence view - or send me an html

**Comment by irina060981** [comment]
>sentence view is supposed to show how many sentences to show around (0 - only current, 1 - one before, one after and so on). Going up works fine, however going down didn't work for me. I was expecting to see the sentences below the current selection but it did only show the word selected and the related alignment. see attachments

It should be >0, 2 - means 2 sentence before and 2 sentence after

**Comment by irina060981** [comment]
> As you can see, show around the sentence does not work properly with the added text.
Oh I see, what you mean - will check

**Comment by monzug** [comment]
> > when I open the html file, it defaults to sentence view. are we ok with this?
> 
> I don't know what item should be the default - I made it as sentence only because it was comfotable for me for developing.
> Could change to anyone else

Uhm, I would rather go with full view

**Comment by monzug** [comment]
> > sentence view: in the translated text, I am missing the edited annotation which is visible in Full view. why is that? see attachment
> 
> Monica, I didn't understand your comment. What is edited annotation?
> Could you attach both - full and sentence view - or send me an html
> 


Irina, it's the text that you can enter before first word or after last word.

**Comment by monzug** [comment]
Irina, not sure if you have seen question 6) about the language. Thanks

**Comment by irina060981** [comment]
>the language is not relevant in html output. Please confirm that this is ok.

You are right - I missed it
What language do you mean? could you show an example?

**Comment by monzug** [comment]
> > the language is not relevant in html output. Please confirm that this is ok.
> 
> You are right - I missed it
> What language do you mean? could you show an example?

In Origin and target language, we can select the language of the text and the translation. It can be saved in json or tsv output and also it's in the filename of the downloaded file. in the case of the html output, I could not see any reference to it. Just asking if this is not in the scope. Thanks

**Comment by irina060981** [comment]
>>>the language is not relevant in html output. Please confirm that this is ok.

>>You are right - I missed it
>>What language do you mean? could you show an example?

>In Origin and target language, we can select the language of the text and the translation. It can be saved in json or tsv output and also it's in the filename of the downloaded file. in the case of the html output, I could not see any reference to it. Just asking if this is not in the scope. Thanks


You are talking about language in the file name, I see. 
I would add it to the file name - there are no requirements for file names.

**Comment by monzug** [comment]
> > > > the language is not relevant in html output. Please confirm that this is ok.
> 
> > > You are right - I missed it
> > > What language do you mean? could you show an example?
> 
> > In Origin and target language, we can select the language of the text and the translation. It can be saved in json or tsv output and also it's in the filename of the downloaded file. in the case of the html output, I could not see any reference to it. Just asking if this is not in the scope. Thanks
> 
> You are talking about language in the file name, I see.
> I would add it to the file name - there are no requirements for file names.

Irina, no need to rush to change anything. it's just a question that is not only about language in filename but also in the text.

**Comment by irina060981** [comment]
>Irina, no need to rush to change anything. it's just a question that is not only about language in filename but also in the text. if there are no requirements about this, we can keep as open questions for next meeting.

I still work on this html output - with some optimzing and updates.
So I could add some changes.

I agree that we should add language somewhere in the interface, I would add - and we could decide on the meeting where it is the best place for that.

**Comment by irina060981** [comment]
>You are talking about language in the file name, I see.
I would add it to the file name - there are no requirements for file names.

- added languages to the file name for HTML output (similiar to other types of download)

>the download as html is possible after alignment. Would it be possible to disable the html option (greyed out or not visible) before tokenization process has started?

done
![image](https://user-images.githubusercontent.com/12377640/107634897-fa59c680-6cb5-11eb-8a21-085619e79f34.png)

![image](https://user-images.githubusercontent.com/12377640/107634918-05acf200-6cb6-11eb-91cd-ee9d00e82255.png)


Also there was a bug with punctuation definition for chinese - fixed

**Comment by irina060981** [comment]
I have found the bug with dividing to sentences for the text that was added after tokenization.
Good catch, @monzug :)
Would fix it

**Comment by monzug** [comment]
Brava!

**Comment by irina060981** [comment]
- added reIndexSentence for tokens edit actions (to fix bug with sentence view)

- fix a bug - for undo/redo delete token

- added language name to html output
![image](https://user-images.githubusercontent.com/12377640/107747463-2e3ef580-6d63-11eb-85c9-da28b1a436e8.png)

**Comment by irina060981** [comment]
I think that I fixed all your notes, @monzug , besides icon for local HTML in Safari.
I spent some time for investigation for icon bug - it is a well-known problem, and I didn't find any proper solution.
And decided not to spend too much time for this for now.

Monica, do you think it is worth a separate issue for future development?

**Comment by monzug** [comment]
yes, a separate issue for the icon. I have to say that I am testing on Big Sur.

**Comment by monzug** [comment]
Irina, I checked the language added in html output. filename is done. the language in the output is redundant. I would not repeat the language for each segment, only once at the top. what do you think?

I added the metadata fields but it does not show in the output. will this feature be implemented in the future?

with the new standalone version, the short view and sentence view are empty.

**Comment by irina060981** [comment]
@monzug , I fixed the bug with short and sentence view - now it works, I am sorry for it - it was a small bug.

>I checked the language added in html output. filename is done, e.g. alignment-html-output-ita-ara-eng.html.
as per the language in the output I think it's redundant. I would not repeat the language for each segment, only once at the top.
see screenshot

I did them for each segment specially for the view when we have several translations
For origin it looks like rather much - but I save the horizontal alignment between origin and target texts.

**Comment by monzug** [comment]
Would it be possible for Origin text to have only one horizontal bar with the language in order to avoid the repetition of the language for each sentence?

**Comment by monzug** [comment]
see also metadata in html output (issue #195 )

**Comment by irina060981** [comment]
>Would it be possible for Origin text to have only one horizontal bar with the language in order to avoid the repetition of the language for each sentence?

It is done for each segment, and one segment could have several sentences.
It would look like this way for full view - 

![image](https://user-images.githubusercontent.com/12377640/108043072-58a6f080-708c-11eb-928d-f46c0920cbb3.png)

Do you like it?

**Comment by irina060981** [comment]
But I think that I could easily remove them from the short and sentence view

**Comment by irina060981** [comment]
Done for short view (3 segments)

![image](https://user-images.githubusercontent.com/12377640/108046765-b9382c80-7090-11eb-951c-14843ac97d03.png)


for sentence view (3 segments)

![image](https://user-images.githubusercontent.com/12377640/108046811-ca813900-7090-11eb-807e-1fbd8b3a8090.png)

**Comment by monzug** [comment]
looks exactly what I meant. will test it later. Thanks

**Comment by monzug** [comment]
Irina, sorry for being such a pain. tested with multiple sentences and actually it doesn't look as good as I thought. see screenshot
![Screen Shot 2021-02-16 at 4 15 03 PM](https://user-images.githubusercontent.com/41396793/108084330-5afd5080-7074-11eb-875e-9079618c0a16.png)

**Comment by irina060981** [comment]
Returned Lang name bar

![image](https://user-images.githubusercontent.com/12377640/108172231-4e4a2c80-7148-11eb-8dc9-7ed8cc409805.png)

![image](https://user-images.githubusercontent.com/12377640/108172266-57d39480-7148-11eb-961f-f8bf1686ba42.png)

![image](https://user-images.githubusercontent.com/12377640/108172298-61f59300-7148-11eb-8c03-9b69e03f28de.png)

**Comment by monzug** [comment]
I'll close this one. it's way too long to follow. and only metadata in origin text is left out. will enter a new issue for that.
